src/test/java/spoon/test/generics/GenericsTest.java

```diff
@@ -633,7 +640,137 @@ public void testGenericTypeReference() throws Exception {
+		assertEquals("A2", sth.adaptType(tpA).getQualifiedName());
+		// in ModelB, "B" is "B2"
+		assertEquals("B2", sth.adaptType(tpB).getQualifiedName());
+		// and so on and so forth
+		assertEquals("C2", sth.adaptType(tpC).getQualifiedName());
+		assertEquals("D2", sth.adaptType(tpD).getQualifiedName());
+
+		CtClass<?> ctModelC = ctModel.filterChildren(new NameFilter<>("ModelC")).first();
+		CtSuperTypeHierarchy sthC = new CtSuperTypeHierarchy(ctModelC);
+		assertEquals("java.lang.Integer", sthC.adaptType(tpA).getQualifiedName());
+		assertEquals("java.lang.RuntimeException", sthC.adaptType(tpB).getQualifiedName());
+		assertEquals("java.lang.IllegalArgumentException", sthC.adaptType(tpC).getQualifiedName());
+		assertEquals("java.util.List", sthC.adaptType(tpD).getQualifiedName());
+	}
 
 
+	@Test
+	public void testCtSuperTypeHierarchy() throws Exception {
+		// contract: a CtSuperTypeHierarchy enables one to perform type resolution of generic types
+		Factory factory = build(new File("src/test/java/spoon/test/generics/testclasses"));
+		CtClass<?> ctClassCelebrationLunch = factory.Class().get(CelebrationLunch.class);
+		CtTypeReference<?> typeReferenceOfDisgust = ctClassCelebrationLunch.filterChildren(new NameFilter<>("disgust")).map((CtTypedElement te)->{
+			return te.getType();
+		}).first();
+		
+		assertEquals("spoon.test.generics.testclasses.CelebrationLunch<java.lang.Integer, java.lang.Long, java.lang.Double>.WeddingLunch<spoon.test.generics.testclasses.Mole>",typeReferenceOfDisgust.toString());
+		//method WeddingLunch#eatMe
+		CtMethod<?> tWeddingLunch_eatMe = typeReferenceOfDisgust.getDeclaration().filterChildren((CtNamedElement e)->"eatMe".equals(e.getSimpleName())).first();
+		
+		CtClass<?> ctClassLunch = factory.Class().get(Lunch.class);
+		//method Lunch#eatMe
+		CtMethod<?> ctClassLunch_eatMe = ctClassLunch.filterChildren((CtNamedElement e)->"eatMe".equals(e.getSimpleName())).first();
+
+		
+		//type of first parameter of  method WeddingLunch#eatMe
+		CtTypeReference<?> ctWeddingLunch_X = tWeddingLunch_eatMe.getParameters().get(0).getType();
+		// X is the type parameter of WeddingLunch
+		assertEquals("X", ctWeddingLunch_X.getSimpleName());
+		//type of first parameter of method Lunch#eatMe
+		CtTypeReference<?> ctClassLunch_A = ctClassLunch_eatMe.getParameters().get(0).getType();
+		assertEquals("A", ctClassLunch_A.getSimpleName());
+		
+		//are these two types same?
+		CtSuperTypeHierarchy typingContextOfDisgust = new CtSuperTypeHierarchy(typeReferenceOfDisgust);
+		
+		// in disgust, X of WeddingLunch is bound to "Model"
+		assertEquals("spoon.test.generics.testclasses.Mole", typingContextOfDisgust.adaptType(ctWeddingLunch_X).getQualifiedName());
+		//adapt A to scope of CelebrationLunch<Integer,Long,Double>.WeddingLunch<Mole>
+
+		// in disgust, the A of Lunch is bound to "Mole"
+		assertEquals("spoon.test.generics.testclasses.Mole", typingContextOfDisgust.adaptType(ctClassLunch_A).getQualifiedName());
```

Comment:
In the hierarchy of WeddingLunch<Mole> the Lunch.A is Mole
In the hierarchy of CelebrationLunch<Integer,Long,Double> the Lunch.A is Double